### PR TITLE
qa: refactor ClassMethodsHydrator and extract two methods

### PR DESCRIPTION
**Pull request by codisart**
Signed-off-by: codisart <louis.celier@gmail.com>

|    Q          |   A
|-------------- | ------
| BC Break      | no
| QA            | yes

### Description
Extract two methods in order to reduce complexity of the extract method of the `ClassMethodsHydrator`

### CODE REVIEW

**Comment by michalbundyra** [comment]
@codisart just a quick question: why do you think these changes might be a BC Break?

Thanks!

File: src/ClassMethodsHydrator.php [hunk 1]
```diff
@@ -194,6 +179,27 @@ public function extract(object $object) : array
         return $values;
     }
 
+    private function initCompositeFilter($object) : Filter\FilterComposite
```
Comment (michalbundyra) [inline src/ClassMethodsHydrator.php]: we can add typehint `object` on `$object` parameter

File: src/ClassMethodsHydrator.php [hunk 2]
```diff
@@ -194,6 +179,27 @@ public function extract(object $object) : array
...
+            return new Filter\FilterComposite(
+                [$object->getFilter()],
+                [new Filter\MethodMatchFilter('getFilter')]
+            );
+        }
+
+        return $this->getCompositeFilter();
+    }
+
+    private function identifyAttributeName($object, $method)
```
Comment (michalbundyra) [inline src/ClassMethodsHydrator.php]: we can add typehints: `object $object, string $method` and RTH: `: string`

**Comment by codisart** [comment]
@michalbundyra I thought about the possibility that some people in userland could have created methods named `initCompositeFilter` or `identifyAttributeName`. And to add these methods would break code using these same methods

**Comment by michalbundyra** [comment]
@codisart 

> I thought about the possibility that some people in userland could have created methods named `initCompositeFilter` or `identifyAttributeName`. And to add these methods would break code using these same methods

Thanks for explanation. This is not considered as a BC break. We can always add any public/protected/private method.

**Comment by froschdesign** [comment]
@codisart 
> I thought about the possibility that some people in userland could have created methods named `initCompositeFilter` or `identifyAttributeName`. And to add these methods would break code using these same methods

This means to kill the concept of minor versions in [semantic versioning](https://semver.org/).

File: src/ClassMethodsHydrator.php (hunk 1 above)
Comment (codisart) [reply src/ClassMethodsHydrator.php]: I just learned that this was added to 7.2. I applied it.

File: src/ClassMethodsHydrator.php (hunk 2 above)
Comment (codisart) [reply src/ClassMethodsHydrator.php]: I need to have this habit. I applied it

**Comment by codisart** [comment]
@froschdesign @michalbundyra  Thank to both of you for the precision. 
I knew that adding code was in the scope of minor versionning.
But I just realized recently that extract a private method could break userland code.
And I was curious to see the POV of public package contributors hence why I used the maybe

**Comment by michalbundyra** [comment]
Thanks, @codisart!

**Comment by weierophinney** [comment]
@codisart — regarding this:

> But I just realized recently that extract a private method could break userland code.
> And I was curious to see the POV of public package contributors hence why I used the maybe

If we considered all the possible method names an extending class could use, we'd never be able to add _anything_ to the public API without it being a BC break! :smile:

Essentially, if we _add_ to the public API, we generally create a new minor version. While minor versions do not break _backwards compatibility_ (i.e., any code you exercise from the public API should continue to work in the same way as it did in the previous version), they may introduce _new features_. Therefore, it is _up to the user to test their code against the new minor version_ to ensure that anything they are doing does not conflict with the new version.

**Comment by codisart** [comment]
@weierophinney Thank you for the answer complement.

**Pull request merged**